### PR TITLE
More LES diagnostics

### LINKED ISSUE #739: high order moments
https://github.com/climate-machine/CLIMA/blob/0ef45f32ef07c4fce847de2d8461144418f8f79e/src/Diagnostics/AtmosDiagnostics.jl#L213

It would save computation to save the HorizontalMean of products of variables and leave the computation of high order moments to offline computation such that the code will save HM(u) , HM(u^2), HM(u^3), HM(u) , HM(u *v), HM(u *v *w) and so on, such that in general the offline computation will read:

vu_second_moment =  HM(u*v) - HM(u)*HM(u)
uvw_third_moment =  HM(u*v*w) - HM(u*w)*HM(v) - HM(v*w)*HM(u) - HM(u*v)*HM(w) + 2*HM(u)*HM(v)*HM(w)

### PR DESCRIPTION
**Pull request by kpamnany**
# Description

- Adds more of the `default` diagnostics from #943 (excluding LWP and cloud cover).
- Changes the variable names for the diagnostics according to the [variable list](https://climate-machine.github.io/CLIMA/latest/DiagnosticVariables/).
- Changes the computation of the vertical fluxes and variances to use density-weighted means.
- Applies the idea from and closes #739.
- Changed `Writers.write_data()` to allow specifying dimension names for a variable.

I have

- [X] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [X] Followed all necessary [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and run `julia .dev/format.jl`

# For review by CLIMA developers

- [X] There are no open pull requests for this already
- [X] CLIMA developers with relevant expertise have been assigned to review this submission
- [X] The code conforms to the [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Diagnostics/atmos_default.jl [hunk 1]
```diff
@@ -194,65 +242,87 @@ function compute_diagnosticsums!(
     horzavgs,
     dsums,
 )
-    zvals = CollectedDiagnostics.zvals
     th = thermo_vars(thermoQ[ijk, e])
     ha = horzavg_vars(horzavgs[Nqk * (ev - 1) + k])
     ds = diagnostic_vars(dsums[Nqk * (ev - 1) + k])
 
     u = state.ρu[1] / state.ρ
+    u′ = ha.ρu / ha.ρ
```
Comment (szy21) [inline src/Diagnostics/atmos_default.jl:250]: The prime symbol is confusing to me because it is commonly used for deviation from the mean in fluid dynamics. @akshaysridhar what do you think?

Comment (kpamnany) [reply src/Diagnostics/atmos_default.jl:250]: How about I rename all `x′` to `avg_x`?

**Comment by kpamnany** [comment]
The file to focus on for code review is `src/Diagnostics/atmos_default.jl`.

More practically, checking out this branch and running BOMEX or one of the other experiments with `--enable-diagnostics` and then checking the output diagnostics to see if they make sense would be very helpful.

**Comment by szy21** [comment]
@yairchn would you like to take a look at this? I may not be able to provide a full review until later this week.

**Comment by kpamnany** [comment]
P.S.: If #958 merges before you get a chance to test, `--enable-diagnostics` should become `--diagnostics default` (or replace 'default' with an interval specification as described there).

File: src/Diagnostics/atmos_default.jl [hunk 2]
```diff
@@ -3,16 +3,70 @@ using ..Atmos: thermo_state, turbulence_tensors
...
+        var_w::FT               # w′w′
+        var_qt::FT              # q_tot′q_tot′
+        var_ei::FT              # e_int′e_int′
+        var_thl::FT             # thl′thl′
+
+        # skewness
+        w3::FT                  # w′w′w′
+
+        # turbulent kinetic energy
+        tke::FT
```
Comment (smarras79) [inline src/Diagnostics/atmos_default.jl:60]: As it is now there is a potential bug in the calculation of tke. 
Since tke can be obtained from other quantities that are written to file, should be just omit it from output?

File: src/Diagnostics/atmos_default.jl [hunk 3]
```diff
@@ -194,65 +241,87 @@ function compute_diagnosticsums!(
...
+        ds.cov_qt_ei += MH * (q_tot - avg_q_tot) * (e_int - avg_e_int)
+        ds.var_qt += MH * (q_tot - avg_q_tot)^2
     end
 
-    # variances
-    ds.uvariance += MH * (u - ũ)^2
-    ds.vvariance += MH * (v - ṽ)^2
-    ds.wvariance += MH * (w - w̃)^2
-
     # skewness
```
Comment (smarras79) [inline src/Diagnostics/atmos_default.jl:320]: #skewness should be replaced with #w3 since they are not the same quantity. Skewness can be obtained from w3 and var_w so that it's not necessary to calculate it here.

File: src/Diagnostics/atmos_default.jl (hunk 2 above)
Comment (smarras79) [inline src/Diagnostics/atmos_default.jl:60]: Nevermind my past comment. TKE is calculated and output correctly.

Comment (akshaysridhar) [reply src/Diagnostics/atmos_default.jl:60]: This was addressed in issue #831

File: src/Diagnostics/atmos_default.jl [hunk 4]
```diff
@@ -194,65 +241,87 @@ function compute_diagnosticsums!(
...
+    ds.thl += MH * avg_θ_liq_ice
+    ds.thv += MH * avg_θ_v
+    ds.ei += MH * avg_e_int
+    ds.hm += MH * avg_h_moi
+    ds.ht += MH * avg_h_tot
+    ds.qt_sgs += MH * avg_qt_sgs
+    ds.ht_sgs += MH * avg_ht_sgs
+
+    # covariances and variances
+    ds.cov_w_rho += MH * (w - avg_w) * (state.ρ - ha.ρ)
```
Comment (szy21) [inline src/Diagnostics/atmos_default.jl:291]: I have talked with @tapios and we want density averaged density here (yes, it's strange). So you would calculate th.ρρ = state.ρ*state.ρ in `compute_thermo`, and avg_ρ = ha.ρρ / ha.ρ, and use avg_ρ instead of ha.ρ here, and only here.

**Comment by szy21** [comment]
We would need density weighted average for the output. For example, `ds.cov_w_qt += MH * (w - avg_w) * (q_tot - avg_q_tot)`, and for the output we would like to have `avg(ds.cov_w_qt*state.ρ)/ha.ρ`, while right now it's `avg(ds.cov_w_qt)`. Maybe the simplest solution is to have another function for density weighted horizontal means?

**Comment by kpamnany** [comment]
> We would need density weighted average for the output.

This isn't clear to me. Let's talk about it.

**Comment by kpamnany** [comment]
Reworking this per discussion with @szy21 and @akshaysridhar so don't worry about reviewing it yet.

File: src/Diagnostics/atmos_core.jl [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    nhorzelem = div(nrealelem, nvertelem)
+
+    if Array ∈ typeof(Q).parameters
+        localvgeo = grid.vgeo
+    else
+        localvgeo = Array(grid.vgeo)
+    end
+
+    @warn """
+Diagnostics ($dgngrp.name): not yet implemented"""
```
Comment (vchuravy) [inline src/Diagnostics/atmos_core.jl:29]: Triple quotes don't need to begin at the beginning of the line.

```
julia>     """
           hello
           """
```
is the same as:
```
julia> """
       hello
       """

**Comment by kpamnany** [comment]
This is ready for review. Please especially look at [this](https://github.com/climate-machine/CLIMA/blob/kp/more-diagnostics/src/Diagnostics/atmos_default.jl#L117-L144) and [this](https://github.com/climate-machine/CLIMA/blob/kp/more-diagnostics/src/Diagnostics/atmos_default.jl#L178-L192).

Also please do run an experiment with these diagnostics enabled and check the output, including cloud fraction, cloud top and cloud base.

**Comment by smarras79** [comment]
> This is ready for review. Please especially look at [this](https://github.com/climate-machine/CLIMA/blob/kp/more-diagnostics/src/Diagnostics/atmos_default.jl#L117-L144) and [this](https://github.com/climate-machine/CLIMA/blob/kp/more-diagnostics/src/Diagnostics/atmos_default.jl#L178-L192). Note that all these (except `avg_rho`) are later divided by `avg_rho`.
> 
> Also please do run an experiment with these diagnostics enabled and check the output, including cloud fraction, cloud top and cloud base.

Fantastic, @kpamnany I'll try it with dycoms this afternoon

**Comment by kpamnany** [comment]
After more discussion with @szy21, @akshaysridhar, @yairchn, and @slifer50, this is being reworked yet again. Please hold on reviewing until I update again tomorrow.

**Comment by kpamnany** [comment]
Only thing remaining is fixing `cld_frac` to be on `z`.

**Comment by szy21** [comment]
@kpamnany Could we have `temp`=temperature too? (not urgent)

File: src/Diagnostics/atmos_default.jl [hunk 6]
```diff
@@ -34,268 +36,261 @@ function atmos_default_init(dgngrp::DiagnosticsGroup, currtime)
...
+    w = state.ρu[3] / state.ρ
+    w′ = w - ha.w
+    e_int′ = thermo.e_int - ha.ei
+    θ_dry′ = thermo.θ_dry - ha.thd
+    θ_vir′ = thermo.θ_vir - ha.thv
+
+    sums.var_u += MH * state.ρu[1] / state.ρ * state.ρu[1]
+    sums.var_v += MH * state.ρu[2] / state.ρ * state.ρu[2]
+    sums.var_w += MH * state.ρu[3] / state.ρ * state.ρu[3]
+    sums.w3 += MH * state.ρu[3] / state.ρ * state.ρu[3] / state.ρ * state.ρu[3]
```
Comment (szy21) [inline src/Diagnostics/atmos_default.jl:219]: Why are these here? (221-224 are the correct definition of var_u, var_v, var_w, and w3.)
